### ISSUE TITLE
Odd behavior with docked sidebar and margin notes

### ISSUE THREAD

**Issue by kdheepak** [quarto-dev/quarto-cli#4719]
### Bug description

With this .quarto.yml file:

```yaml
project:
  type: website

website:
  title: "blog"
  sidebar:
    style: docked

format:
  html:
    theme:
      - default
```

This index.qmd file:

````md
---
title: test quarto docked sidebar with margin notes
date: 2023-02-25T21:14:15-0500
---

## title for toc

[Here's a aside note.]{.aside}

````

renders this:

<img width="1512" alt="Screenshot 2023-03-08 at 23 28 53" src="https://user-images.githubusercontent.com/1813121/223917463-3e01b261-9e9f-4196-8b9e-830c8b6ce505.png">

But this index.qmd file:

````md
---
title: test quarto docked sidebar with margin notes
date: 2023-02-25T21:14:15-0500
---

## title for toc

<!-- [Here's a aside note.]{.aside} -->

````

renders this:

<img width="1512" alt="Screenshot 2023-03-08 at 23 29 10" src="https://user-images.githubusercontent.com/1813121/223917508-5edb516c-8dad-4cac-9b6e-8fd8f53a53a7.png">

In this particular theme, seemingly the only difference here is the horizontal rule below the header, but basically, when there is a margin note _with_ `sidebar: style: docked`, there's lots of excess space on the right, and the center body seems to be shrunk. When you remove the margin note, the layout appears to change.

With `sidebar: style: float`, adding or removing a margin note does not affect the layout.

I like the behavior of the docked sidebar, but when I write blog posts with margin notes the body shrinks.

1) It doesn't match the layout of other blog posts without margin notes
2) it makes the body smaller and makes the page longer than it would be otherwise
3) Since this doesn't happen with `sidebar: style: float`, I'm assuming this is a bug.

```
shell> quarto --version
1.3.242
```

```
Platform Info:
  OS: macOS (arm64-apple-darwin21.5.0)
  CPU: 10 × Apple M1 Pro
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 8 virtual cores
```

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [ ] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.

**Comment by dragonstyle** [comment]
I'm not able to reproduce what you're seeing. When I add and remove margin notes, the layout is indeed adjusted, as the grid system alters the page layout to create space for margin elements. This is by design, and in my testing does so consistently between floating and docked style layouts. I followed the above reproducible steps and ended up with a document with a body with of 697 pixels width for both floating and docked styles.

**Comment by kdheepak** [comment]
Thanks for reopening! Let me know if I can provide any additional information.

**Comment by dragonstyle** [comment]
The only explanation that I have for this would be if you had set `page-layout: full` somewhere. that layout specifically means to use the extra space in the right sidebar when margin elements aren't present, so it intentionally grows the body content region. If you omit this setting, I am unable to cause the body grid to change substantially.

Is this setting declared in your `_quarto.yml` or `_metadata.yml` file somewhere? If you omit this, it should result in a more consistent appearance across documents.

**Comment by kdheepak** [comment]
I don't have `page-layout: full` anywhere.

But I did a bisect on all the options in my `.quarto.yml` file. 

I think there is also another related to the `navbar` for some reason.

Here's the bare minimal `quarto.yml`:

```
project:
  type: website

website:
  title: "blog"
  sidebar:
    style: docked
  navbar: true
format:
  html:
    theme:
      - default # or journal 
```

With this markdown file:

````md
---
title: testing
date: 2023-02-25T21:14:15-0500
---

I would like to present a mental model for understanding the basics of variable assignment and mutability in [Julia](https://julialang.org) using an analogy involving books and bookshelves, from a complete beginner's perspective.

[While analogies can be useful initially to understand complex concepts quickly by making comparisons to something familiar, they can also be oversimplified, leading to an incomplete understanding of the topic. Once you are comfortable, I recommend diving deeper into the subject matter directly.]{.aside}
````

With and without `navbar: true` I get very different `body` `width`. That's one issue.

--- 

The second issue here is what my original post was about which I'm able to reproduce now when I use `toc: true` AND `toc-location: left` AND use `aside` notes.. This works as expected when using `toc-location: right`. 

With this quarto file:

```
project:
  type: website

author: "Dheepak Krishnamurthy"

website:
  title: "blog"
  sidebar:
    alignment: left
    style: docked
format:
  html:
    theme:
      - journal
    toc: true
    toc-location: left

```


Use this markdown file:

````md
---
title: testing
date: 2023-02-25T21:14:15-0500
---

I would like to present a mental model for understanding the basics of variable assignment and mutability in [Julia](https://julialang.org) using an analogy involving books and bookshelves, from a complete beginner's perspective.

[While analogies can be useful initially to understand complex concepts quickly by making comparisons to something familiar, they can also be oversimplified, leading to an incomplete understanding of the topic. Once you are comfortable, I recommend diving deeper into the subject matter directly.]{.aside}

# Assignment

In Julia, the `=` operator can be used to create a new label.
```` 

The width of the body is much shorter than when I try to render without the `aside` notes.

**Comment by dragonstyle** [comment]
1) `navbar: true` 

This is a great observation - really good detective work. What is really happening as you enable and disable the navbar (with sidebar set to true so it always appear, even when completely empty), content like the title and the search UI are moving down to the sidebar. We are seeing those elements (children) in the sidebar when we render the page and making room for them by slimming the content (e.g. adding that `slimcontent` class). I'm 100% certain what the correct behavior is in this situation, what we are doing makes sense I think, but perhaps there is something better we could be doing. When you set navbar to true, the elements shift up into the navbar, the sidebar has no children, and so we don't `slim` the content.

2) The second scenario is actually another version of the same thing. Adding content to the right margin  shifts the grid system into the slim content mode since there is now content in both the sidebar and the margin.

One thing I could think of doing is adding some options to control the heuristics of the grid system (e.g. disable some of the heuristics or force it into particular modes) to better cope with scenarios where the heuristics are providing an undesirable outcome. If you have other ideas, I'm very open to suggestions here too...

**Comment by kdheepak** [comment]
I'm not sure what the best solution here would be as I don't fully understand the quarto control flow and I'm not a web developer by any means, I've just dabbled in it. 

Here's my understanding of what currently is happening.

1) Markdown files are read by quarto 
2) Quarto preprocesses the markdown files (?)
3) Quarto converts the preprocessed markdown files to ipynb if required, executes the file, converts it back to markdown
4) Quarto passes the preprocessed markdown content to Pandoc with various filters
5) Quarto gets HTML output and applies a number of css classes based on whether there's a navbar, whether there's a toc, whether there's references, etc.

Feel free to correct me if I'm mistaken.

The last step is probably the only aspect that matters for this issue. 

Ideally, I'd like to see the exact same HTML file being generated with different css files being generated, instead of css classes being added or removed. My understand was that with CSS grid, it isn't necessary to add various different classes, and the different CSS can be applied to support completely different layouts. 

Maybe this process of adding CSS classes is a byproduct of using Bootstrap? Maybe I should try to write a `none` theme or a `pandoc` theme and write my own CSS file for that to understand what is going on here.

**Comment by dragonstyle** [comment]
Well, to clarify, what is really happening is that we're altering the grid sizes / geometry to accommodate the content that we believe is being rendered on the page. The heuristics are basically doing their best to optimize the grid, but this will cause variance as different pages will have different 'optimized' grid sizes based upon the types of content that are present on that page.

I did add a new option which will allow you to specify the grid content behavior - like so:

```yaml
format:
  html:
    grid:
      content-mode: standard
```

This will force the page to use a standard grid sizing no matter what content is present (e.g. it will not grow the margin to allow room for margin content, it will just place the margin content in the standard size margin). 

I just started a fresh pre-release build which contains this new option (should be ready in 10-15 minutes), LMK if this addresses what you're going for.
